rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java

```diff
@@ -384,37 +389,74 @@ else if (isCached.get(idx).getLocation() == Location.CACHED) {
             else {
                 if (isCached.get(idx).getLocation() == Location.NON_LOCAL) {
                     String remoteLocation = isCached.get(idx).getRemoteLocation();
-                    log.debug(String.format("Sending block %d to NonLocalReadRequestChain to node : %s", blockNum, remoteLocation));
-                    if (!nonLocalRequests.containsKey(remoteLocation)) {
-                        NonLocalReadRequestChain nonLocalReadRequestChain =
-                            new NonLocalReadRequestChain(remoteLocation, fileSize, lastModified, conf,
-                                remoteFileSystem, remotePath, clusterType.ordinal(), strictMode, statistics);
-                        nonLocalRequests.put(remoteLocation, nonLocalReadRequestChain);
+
```

Comment:
This is a huge function. Refactor is probably out of scope of this PR. Is there a way we can avoid adding to the number of things it does ? I'll try to understand what is going on here to help break it up.